### PR TITLE
Substring visit matching.

### PR DESCRIPTION
**Pull request by rnewman**
@bgrins!

### CODE REVIEW

File: app/services/storage.js [hunk 1]
```diff
@@ -180,10 +180,37 @@ export class ProfileStorage {
...
+      out.push(row.url);
+    }
+    return out;
+  }
 
+  async visitedMatches(substring, since = 0, limit = 10) {
+    // Fetch all places visited, with the latest timestamp for each.
+    // We can't limit here, because we don't know matching URLs.
+    // TODO: inverting this query might work better if `since` is long ago.
+    const sub = `SELECT place, MAX(ts) AS latest
```
Comment (ncalexan) [inline app/services/storage.js]: There's a good deal of duplication here.  Can we streamline?

Perhaps this mostly falls away when we start maintaining our in-memory views?

**Comment by ncalexan** [comment]
These look good.

File: app/services/storage.js (hunk 1 above)
Comment (rnewman) [inline app/services/storage.js]: This'll get reworked a little when we have to string match on more things, and yes, there will be some intersection of in-DB views, in-DB materialized views, and in-memory materialized views.

File: app/services/storage.js [hunk 2]
```diff
@@ -78,7 +84,51 @@ export class ProfileStorage {
...
+    return this;
+  }
+
+  /**
+   * Begins a transaction and runs `f`. If `f` returns a promise that rejects,
+   * rolls back the transaction and passes through the rejection. If `f` resolves
+   * to a value, commits the transaction and passes through the resolved value.
+   *
+   * @param f the function to call within a transaction.
+   * @returns {*} the result or rejection of `f`.
```
Comment (bgrins) [inline app/services/storage.js]: All of these comments say returns `{*}` but really they return a `Promise` that resolves with `{*}`

File: app/services/storage.js [hunk 3]
```diff
@@ -78,7 +84,51 @@ export class ProfileStorage {
...
+   */
+  inTransaction(f) {
+    return this.db
+               .run('BEGIN TRANSACTION')
+               .then(f)
+               .then((result) =>
+                 this.db
+                     .run('COMMIT')
+                     .then(() => Promise.resolve(result)))
+               .catch((err) =>
```
Comment (bgrins) [inline app/services/storage.js]: Is there a good way to add a test case for failures on functions in this file? It can be hard to follow all of the promise rejections / catching to figure out what the expected return values are in the case of a failure
